FormBuilder - Now you can make Tab Sets 🎉

colemanw opened the pull request:
Overview
----------------------------------------
Build tabsets with the FormBuilder drag-n-drop editor!

![image](https://github.com/user-attachments/assets/e948820c-1c1e-4bfa-9a0c-14601144df32)

Technical Details
-------
This breaks with the old `crmUiTabSet` jQuery-Ang-hybrid directive and re-implements tabs with pure AngularJs.

colemanw committed:
FormBuilder - add tabset element

Allows a tabset element to be configured in FormBuilder

colemanw committed:
FormBuilder - code readability cleanup

Same functionality, but a bit more clear what's going on

colemanw committed:
FormBuilder - support counts in tabHeader

This was a challenge! I had to rework the afTabset & afTab components,
because there's no good way to access
properties of a controller in transcluded html.

Using transclude and then accessing `$parent.count` worked, but I'm not sure why.
Angular scope is dark magic...

mattwire commented:
@nicol @artfulrobot is this good from a theming point of view?

colemanw commented:
> @nicol @artfulrobot is this good from a theming point of view?

I think there are 2 lines of css in this PR to port to riverLea. Also significant changes to the markup of the "Campaign Dashboard" screen, but it's all BS3-compliant markup.

vingle commented on ext/afform/core/ang/afCore.css:
```diff
@@ -82,3 +82,9 @@ af-form {
+/* Tabs */
+#bootstrap-theme .af-tabpanel {
+  display: block;
+  padding: 10px;
```
Thanks for flagging, I didn't get this – as confusingly I'm 'nicol' on gitlab and vingle here.

@artfulrobot - all that matters for now is RiverLea adds this line, and swaps it for rems, then (`var(--crm-m2)` = 10px). Too many pixels in core to sweat.

If this is a normally padding instance for something with a border, then `--crm-padding-reg` (16px) or `--crm-padding-inset` (8px) might be a better variable as it lets Streams can adjust all instances at once.

colemanw replied:
There's nothing precise about the 10px here, it just gives a bit of padding so the tab content doesn't look squashed against the edge. Feel free to use a value more consistent with other styles.

vingle commented:
Have had a quick test of the Tab Building UI - works nice and intuitively and seems to work with RiverLea well (prob thanks to the BS3 markup).

That said when I try to view my test tabbed form - it doesn't seem to display - not sure if that's a config shortfall on my side or an actual bug.

colemanw commented:
@vingle I think the only problem is you gave your form the same url as another form. I changed it to something else and now it works: http://core-31944-7exld.test-1.civicrm.org:8003/civicrm/quick-test

vingle commented:
@colemanw - ah thanks. 

The tabs themselves work nicely but it doesn't pick up the wrapper. Is there any reason you didn't re-use BS3's `panel` wrapper with a tab pane using `panel-body`? As that's the pattern in API4 explorer and FB editor - so it would then inherit the rest of the theme.

vingle commented:
(tho, tbf, there's a bunch of different tab markup classes used in different bits of civi - ie searchkit listings use a slightly different pattern)

colemanw commented:
> @vingle The tabs themselves work nicely but it doesn't pick up the wrapper. Is there any reason you didn't re-use BS3's `panel` wrapper with a tab pane using `panel-body`? As that's the pattern in API4 explorer and FB editor - so it would then inherit the rest of the theme.

I think BS3 panels and tabs are different widgets. As you note, some screens in Civi combine them to get a tabbed panel, but they can also be used separately.

vingle commented:
I guess my point is more that according to ThemeTest Civi currently has 7 patterns for making tabs (https://lab.civicrm.org/extensions/themetest/-/tree/main/ang/themetest/snippets?ref_type=heads) - does it need an eighth? (and if it does, could this be the best / final pattern, similar to how we did with .crm-accordion?)

colemanw commented:
@vingle I guess I didn't see this as inventing a new pattern but rather re-using the BS3 standard. FormBuilder tends to follow BS3 markup conventions.

vingle commented:
@colemanw the tab buttons themselves use BS3, that's why they look like tabs in Greenwich (and RiverLea):

<img width="244" alt="image" src="https://github.com/user-attachments/assets/c89f95c6-2d52-439d-918b-0de8c07c2700" />

But the wrapper around the tabs has no class and the tab pane has a new class `.af-tabpanel`. If you don't want to use `.panel` & `.panel-body` as some bits of Civi use to define tab regions, could we add a new class, ie `crm-tabs` and `crm-tabs-pane`? It would I think have a low impact to add this to the markup, but at least then there would be clear classes to target to put, say, a border/drop-shadow around the entire tab block, and padding/bg colour on the tab pane itself..

FWIW, I appreciate even BS5 doesn't create tab wrappers (https://getbootstrap.com/docs/5.1/components/navs-tabs/) yet in Civi all the tab instances have a bounding border, there isn't a pure nav tab style that Bootstrap references, it's always in a box.

<img width="663" alt="image" src="https://github.com/user-attachments/assets/ee415fc8-645b-4068-95cb-1a574001e808" />


When it comes to listings of SearchKit and FormBuilder the lack of this bounding wrapper class creates a bit of a headache to target for consistent themeing… ie

Eg. to create the border around the tab pane below I'm doing some really specific targeting, that then has to be repeated with different specific targets for the FormBuilder listing.. it would save a lot of effort if there was just a consistent `crm-tab` type class on the wrapper.

<img width="1201" alt="image" src="https://github.com/user-attachments/assets/543173ea-911f-4b81-bcb0-87b419caa457" />

ufundo commented:
Bit of a curveball, but saw something about adding a new tab implementation.

If doing that and want it to be *the* tab implementation, then it would be cool to consider  leapfrogging AngularJS and using a WebComponent. I've been experimenting and seems to work pretty seamlessly across  AngularJS / Smarty templates and has the beauty  of being 50 lines of JS with no dependencies...

Can try a PR tomorrow..

colemanw commented:
@ufundo sounds good to me. Implementing the tabs in ang was only a few lines of code so not much to port.

colemanw commented:
@ufundo honestly the hardest thing about this PR was getting the count to update in the tabHeader. It's difficult in AngJS to share a scope variable between a directive and its transcluded content. But I got it working so that:

1. In formBuilder you can select which searchDisplay within a tab you want to be linked to the tab-count.
2. When viewing the form, inactive tabs will run a small background request to fetch only the count, but won't fetch the full result set until the tab is active.

This might be trickier to do with a web component? How can you pass angularJS scope variables to a web component?

ufundo commented:
Hey @colemanw here's where I got to with it today: https://github.com/civicrm/civicrm-core/pull/32038

I'd say it's an interesting experiment but probably not one to wait on longer for your PR here (unless it inspires you to have a go working through the outstanding issues)

colemanw committed:
Afform - standardize tab css classes

colemanw commented:
Ok I've taken @vingle's suggestions and updated the css classes to use something more standard. Now it's:

- `crm-tabset` (pre-existing class, for the wrapper)
- `crm-tab-list` (new class for the ul)
- `crm-tab-button` (pre-existing class for the li)
- `crm-tab-panel` (new class for the tab content wrapper)

I've updated the other angular-based tabs to use these classes consistently. We could also add them to our smarty-based tabs.

vingle commented:
That's great. I'll add support in RiverLea and a snippet in ThemeTest for testing / potentially form the basis for changing smarty/etc tabs.
